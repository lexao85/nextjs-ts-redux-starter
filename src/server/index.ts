import * as compress from 'compression';
import * as express from 'express';
import * as next from 'next';
import * as spdy from 'spdy';
import * as lruCache from 'lru-cache';
import * as fs from 'fs';
import * as geoipLite from 'geoip-lite';
import { parse } from 'url';
import { IncomingMessage, ServerResponse } from 'http';

import serverConfig from './server.config';

const port = parseInt(process.env.PORT, 10) || 3000;
const dev = process.env.NODE_ENV !== 'production';
const app = next({ dev, dir: './src' });
const handle = app.getRequestHandler();

const ssrCache = lruCache({
  max: 100,
  maxAge: 1000 * 60 * 60, // 1hour
});

const config = serverConfig(dev);

app.prepare()
  .then(() => {
    const server = express();

    if (!!config.useTextCompression) {
      server.use(compress());
    }

    server.get('*', (req, res) => {
      const parsedUrl = parse(req.url, true);
      const { pathname, query } = parsedUrl;

      if (config.cacheMap && config.cacheMap[pathname]) {
        renderAndCache(req, res, config.cacheMap[pathname], query);
      } else if (pathname === '/clear_cache') {
        ssrCache.reset();
        res.write('cache cleared');
        res.end();
      } else if (pathname === '/get_geoip_info') {
        const geoipRes = geoipLite.lookup(req.ip);
        res.write(`ip: ${req.ip}\n`);
        res.write(geoipRes ? JSON.stringify(geoipRes) : 'nothing found');
        res.end();
      } else {
        handle(req, res, parsedUrl);
      }
    });

    if (config.certificateSettings) {
      process.env.CONNECTION_TYPE = 'https';
      const { crtPath, keyPath } = config.certificateSettings;
      spdy.createServer(
        {
          cert: fs.readFileSync(crtPath),
          key: fs.readFileSync(keyPath),
        },
        server)
        .listen(port, (err) => {
          if (err) {
            throw new Error(err);
          }

          console.log(`> Ready on https://0.0.0.0:${port}`);
        });
    } else {
      server.listen(port, (err) => {
        if (err) {
          throw err;
        }
        console.log(`> Ready on http://localhost:${port}`);
      });
    }
  });

function getCacheKey(req: IncomingMessage) {
  return `${req.url}`;
}

// tslint:disable-next-line:max-line-length
async function renderAndCache(req: IncomingMessage, res: ServerResponse, pagePath: string, queryParams?: any) {
  const key = getCacheKey(req);

  // If we have a page in the cache, let's serve it
  if (ssrCache.has(key)) {
    res.setHeader('x-cache', 'HIT');
    res.setHeader('Content-Type', 'text/html');
    res.end(ssrCache.get(key));
    return;
  }

  try {
    // If not let's render the page into HTML
    const html = await app.renderToHTML(req, res, pagePath, queryParams);

    // Something is wrong with the request, let's skip the cache
    if (res.statusCode !== 200) {
      res.end(html);
      return;
    }

    // Let's cache this page
    ssrCache.set(key, html);

    res.setHeader('x-cache', 'MISS');
    res.setHeader('Content-Type', 'text/html');
    res.end(html);
  } catch (err) {
    app.renderError(err, req, res, pagePath, queryParams);
  }
}
