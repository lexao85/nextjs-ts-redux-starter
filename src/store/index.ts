import { AppState } from './ducks/app';
import { CartState } from './ducks/cart';
import { applyMiddleware, createStore } from 'redux';
import { composeWithDevTools } from 'redux-devtools-extension';
import reduxThunk from 'redux-thunk';
import ducks from './ducks';

export interface IStore {
  app: AppState;
  cart: CartState;
}

export function initializeStore(initialState?: IStore) {
  return createStore(ducks, initialState, composeWithDevTools(applyMiddleware(reduxThunk)));
}
