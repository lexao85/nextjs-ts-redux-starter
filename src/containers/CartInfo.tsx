import { connect } from 'react-redux';
import React from 'react';
import { IStore } from '../store';

interface IStateProps {
  totalCount: number;
  totalPrice: number;
  currency: string;
}

// tslint:disable-next-line:function-name
export function CartInfo(props: IStateProps) {
  return (
    <div style={{ padding: 5, pointerEvents: 'none' }} >
      <div>{`Elements in cart: ${props.totalCount}`}</div>
      <div>{`Total price: ${props.currency}${props.totalPrice.toFixed(2)}`}</div>
    </div>
  );
}

const mapStateToProps = (state: IStore): IStateProps => {
  const { totalCount, totalPrice } = state.cart;
  return { totalCount, totalPrice, currency: state.app.currency };
};

export default connect(mapStateToProps)(CartInfo);
