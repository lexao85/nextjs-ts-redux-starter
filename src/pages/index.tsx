import Head from 'next/head';
import Link from 'next/link';
import React from 'react';
import ProductList from '../containers/ProductList';
import CartInfo from '../containers/CartInfo';

export default () => {
  return (
    <div>
      <Head key="title">
        <title>Main</title>
      </Head>
      <div style={{ width: '100%', display: 'inline-block' }}>
        <Link href="/cart" as="/cart">
          <div style={{ float: 'right', cursor: 'pointer', pointerEvents: 'all' }}>
            <CartInfo />
          </div>
        </Link>
      </div>
      <ProductList />
    </div>
  );
};
