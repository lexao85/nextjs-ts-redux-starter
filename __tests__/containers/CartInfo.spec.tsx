import React from 'react';
import { Provider } from 'react-redux';
import { mount, shallow } from 'enzyme';
import CartInfo, { CartInfo as DumbCartInfo } from '../../src/containers/CartInfo';
import reduxMockStore from 'redux-mock-store';
import { IStore } from 'src/store';

describe('>>> CartInfo --- Snapshot', () => {
  it('+++capturing Snapshot of CartInfo', () => {
    const rendered = shallow(<DumbCartInfo currency="$" totalCount={5} totalPrice={103.78} />);
    expect(rendered).toMatchSnapshot();
  });
});

describe('>>> CartInfo --- Render with Store', () => {
  const mockStore = reduxMockStore<IStore>();
  const initialStore: IStore = {
    app: {
      currency: '$',
      productList: [],
    },
    cart: {
      items: [],
      totalCount: 0,
      totalPrice: 0,
    },
  };

  it('+++ render the connected CartInfo', () => {
    const store = mockStore(initialStore);
    const wrapper = mount(<Provider store={store}><CartInfo /></Provider>);
    expect(wrapper.length).toEqual(1);
  });

  it('+++ correct elements count in cart info', () => {
    const store = mockStore(initialStore);
    const wrapper = mount(<Provider store={store}><CartInfo /></Provider>);
    expect(wrapper.find('div[children="Elements in cart: 0"]').length).toEqual(1);
  });
});
