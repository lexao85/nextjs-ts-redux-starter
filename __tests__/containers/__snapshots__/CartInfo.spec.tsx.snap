// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`>>> CartInfo --- Snapshot +++capturing Snapshot of CartInfo 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <CartInfo
    currency="$"
    totalCount={5}
    totalPrice={103.78}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div>
          Elements in cart: 5
        </div>,
        <div>
          Total price: $103.78
        </div>,
      ],
      "style": Object {
        "padding": 5,
        "pointerEvents": "none",
      },
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Elements in cart: 5",
        },
        "ref": null,
        "rendered": "Elements in cart: 5",
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Total price: $103.78",
        },
        "ref": null,
        "rendered": "Total price: $103.78",
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div>
            Elements in cart: 5
          </div>,
          <div>
            Total price: $103.78
          </div>,
        ],
        "style": Object {
          "padding": 5,
          "pointerEvents": "none",
        },
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Elements in cart: 5",
          },
          "ref": null,
          "rendered": "Elements in cart: 5",
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Total price: $103.78",
          },
          "ref": null,
          "rendered": "Total price: $103.78",
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
